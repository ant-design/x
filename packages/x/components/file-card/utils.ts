export function getSize(size: number) {
  let retSize = size;
  const units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB'];
  let unitIndex = 0;

  while (retSize >= 1024 && unitIndex < units.length - 1) {
    retSize /= 1024;
    unitIndex++;
  }

  return `${retSize.toFixed(0)} ${units[unitIndex]}`;
}

export function matchExt(suffix: string, ext: string[]) {
  return ext.some((e) => suffix.toLowerCase() === `.${e}`);
}
