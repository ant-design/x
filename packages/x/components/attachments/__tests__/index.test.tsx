import React from 'react';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import { fireEvent, render, waitFakeTimer } from '../../../tests/utils';
import Attachments, { type AttachmentsProps } from '..';

describe('attachments', () => {
  mountTest(() => <Attachments />);
  rtlTest(() => <Attachments />);
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  const mockItems = Array.from({ length: 5 }).map(
    (_, index) =>
      ({
        uid: String(index),
        name: `file-${index}.jpg`,
        status: 'done',
        thumbUrl: 'https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png',
        url: 'https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png',
      }) as const,
  );

  const renderAttachments = (props?: AttachmentsProps) => {
    return <Attachments beforeUpload={() => false} {...props} />;
  };

  it('add and remove file', async () => {
    const onChange = jest.fn();
    const { container } = render(
      renderAttachments({
        onChange,
      }),
    );

    // Add file
    fireEvent.change(container.querySelector('input')!, {
      target: { files: [{ file: 'foo.png' }] },
    });
    await waitFakeTimer();
    expect(onChange.mock.calls[0][0].fileList).toHaveLength(1);
    onChange.mockClear();

    // Remove file
    fireEvent.click(container.querySelector('.ant-file-card-list-remove')!);
    await waitFakeTimer();
    expect(onChange.mock.calls[0][0].fileList).toHaveLength(0);
  });
  it('support classnames and styles', () => {
    render(
      renderAttachments({
        styles: {
          placeholder: {
            color: 'blue',
          },
          upload: {
            color: 'red',
          },
        },
        classNames: {
          placeholder: 'placeholder',
          upload: 'upload',
        },
        items: mockItems,
      }),
    );
  });
  it('add and remove file but can stop remove', async () => {
    const onChange = jest.fn();
    const { container } = render(
      renderAttachments({
        onChange,
        onRemove: () => false,
      }),
    );

    // Add file
    fireEvent.change(container.querySelector('input')!, {
      target: { files: [{ file: 'foo.png' }] },
    });
    await waitFakeTimer();
    expect(onChange.mock.calls[0][0].fileList).toHaveLength(1);
    onChange.mockClear();

    // Remove file
    fireEvent.click(container.querySelector('.ant-file-card-list-remove')!);
    await waitFakeTimer();
    expect(container.querySelectorAll('.ant-file-card-list-item')).toHaveLength(1);
    expect(onChange).not.toHaveBeenCalled();
  });

  it('overflow: scrollX', () => {
    const { container } = render(
      renderAttachments({
        overflow: 'scrollX',
        items: mockItems,
      }),
    );

    expect(container.querySelector('.ant-file-card-list-overflow-scrollX')).toBeTruthy();
  });

  it('overflow: scrollY', () => {
    const { container } = render(
      renderAttachments({
        overflow: 'scrollY',
        items: mockItems,
      }),
    );

    expect(container.querySelector('.ant-file-card-list-overflow-scrollY')).toBeTruthy();
  });

  it('card list description done', () => {
    const { container } = render(
      renderAttachments({
        items: [
          {
            uid: '1',
            name: 'file-1.txt',
            status: 'done',
            description: 'test description',
          },
        ],
      }),
    );

    expect(container.querySelector('.ant-file-card-file-description')?.textContent).toBe(
      'test description',
    );
  });

  it('card list description uploading', () => {
    const { container } = render(
      renderAttachments({
        items: [
          {
            uid: '2',
            name: 'file-2.txt',
            status: 'uploading',
            percent: 50,
          },
        ],
      }),
    );

    expect(container.querySelector('.ant-file-card-file-description')?.textContent).toBe('50%');
  });

  it('card list description error', () => {
    const { container } = render(
      renderAttachments({
        items: [
          {
            uid: '3',
            name: 'file-3.txt',
            status: 'error',
            response: 'Error message',
          },
        ],
      }),
    );

    expect(container.querySelector('.ant-file-card-file-description')?.textContent).toBe(
      'Error message',
    );
  });

  it('image list mask', () => {
    const { container } = render(
      renderAttachments({
        items: [
          {
            uid: '1',
            name: 'image uploading preview.png',
            status: 'uploading',
            percent: 33,
            thumbUrl:
              'https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png',
            url: 'https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png',
          },
          {
            uid: '2',
            name: 'image error preview.png',
            status: 'error',
            response: 'Server Error 500',
            thumbUrl:
              'https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png',
            url: 'https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png',
          },
        ],
      }),
    );

    expect(container.querySelector('.ant-attachment-list-card-file-img-mask')).toBeTruthy();
    expect(container.querySelector('.ant-progress')).toBeTruthy();
    expect(container.querySelector('.ant-attachment-list-card-ellipsis')?.textContent).toBe(
      'Server Error 500',
    );
  });

  it('maxCount', async () => {
    const onChange = jest.fn();
    const presetFiles = Array.from({ length: 5 }).map(
      (_, index) =>
        ({
          uid: String(index),
          name: `file-${index}.jpg`,
          status: 'done',
          thumbUrl: 'https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png',
          url: 'https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png',
        }) as const,
    );

    const { container } = render(
      renderAttachments({
        maxCount: 5,
        items: presetFiles,
        onChange,
      }),
    );

    expect(container.querySelectorAll('.ant-file-card-list-item')).toHaveLength(5);

    const uploadBtn = container.querySelector('.ant-upload-wrapper .ant-btn');
    expect(uploadBtn).toBeTruthy();

    if (uploadBtn) {
      fireEvent.click(uploadBtn);
      const fileInput = container.querySelector('input[type="file"]');
      if (fileInput) {
        fireEvent.change(fileInput, {
          target: { files: [new File(['test'], 'test-file.jpg', { type: 'image/jpeg' })] },
        });
        await waitFakeTimer();
        if (onChange.mock.calls.length > 0) {
          const lastCall = onChange.mock.calls[onChange.mock.calls.length - 1];
          expect(lastCall[0].fileList.length).toBeLessThanOrEqual(5);
        }
      }
    }
  });
});
