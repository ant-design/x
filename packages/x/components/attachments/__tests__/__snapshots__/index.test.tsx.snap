// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`attachments rtl render component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-attachment css-var-root ant-attachment-rtl"
  dir="rtl"
>
  <div
    class="ant-file-card-list-wrapper"
  >
    <div
      class="ant-file-card-list ant-attachment-list  css-var-root ant-file-card-rtl"
      style="display: none;"
    >
      <span
        class="ant-upload-wrapper css-var-root ant-upload-rtl"
      >
        <div
          class="ant-upload ant-upload-select"
        >
          <span
            class="ant-upload"
          >
            <input
              accept=""
              name="file"
              style="display: none;"
              type="file"
            />
            <button
              class="ant-btn css-var-root ant-btn-dashed ant-btn-color-default ant-btn-variant-dashed ant-btn-rtl ant-attachment-list-upload-btn"
              type="button"
            >
              <span
                aria-label="plus"
                class="anticon anticon-plus ant-attachment-list-upload-btn-icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="plus"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                  />
                  <path
                    d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </div>
      </span>
    </div>
  </div>
  <div
    class="ant-attachment-placeholder"
  >
    <span
      class="ant-upload-wrapper css-var-root ant-upload-rtl"
    >
      <div
        class="ant-upload ant-upload-drag ant-upload-rtl"
        style="padding: 0px; border: 0px; background: transparent;"
      >
        <span
          class="ant-upload ant-upload-btn"
          role="button"
          tabindex="0"
        >
          <input
            accept=""
            name="file"
            style="display: none;"
            type="file"
          />
          <div
            class="ant-upload-drag-container"
          >
            <div
              class="ant-attachment-placeholder-inner ant-flex css-var-root ant-flex-align-center ant-flex-justify-center ant-flex-rtl ant-flex-vertical"
            >
              <span
                class="ant-typography ant-typography-rtl ant-attachment-placeholder-icon css-var-root"
              />
              <h5
                class="ant-typography ant-typography-rtl ant-attachment-placeholder-title css-var-root"
              />
              <span
                class="ant-typography ant-typography-rtl ant-typography-secondary ant-attachment-placeholder-description css-var-root"
              />
            </div>
          </div>
        </span>
      </div>
    </span>
  </div>
  <div
    class="ant-attachment-drop-area css-var-root"
    style="display: none;"
  >
    <div
      class="ant-attachment-placeholder"
    >
      <span
        class="ant-upload-wrapper css-var-root ant-upload-rtl"
      >
        <div
          class="ant-upload ant-upload-drag ant-upload-rtl"
          style="padding: 0px; border: 0px; background: transparent;"
        >
          <span
            class="ant-upload ant-upload-btn"
            role="button"
            tabindex="0"
          >
            <input
              accept=""
              name="file"
              style="display: none;"
              type="file"
            />
            <div
              class="ant-upload-drag-container"
            >
              <div
                class="ant-attachment-placeholder-inner ant-flex css-var-root ant-flex-align-center ant-flex-justify-center ant-flex-rtl ant-flex-vertical"
              >
                <span
                  class="ant-typography ant-typography-rtl ant-attachment-placeholder-icon css-var-root"
                />
                <h5
                  class="ant-typography ant-typography-rtl ant-attachment-placeholder-title css-var-root"
                />
                <span
                  class="ant-typography ant-typography-rtl ant-typography-secondary ant-attachment-placeholder-description css-var-root"
                />
              </div>
            </div>
          </span>
        </div>
      </span>
    </div>
  </div>
</div>
`;
